source/distributed/cell_weights.cc

```diff
@@ -63,71 +44,186 @@ namespace parallel
 
   template <int dim, int spacedim>
   void
-  CellWeights<dim, spacedim>::register_constant_weighting(
-    const unsigned int factor)
+  CellWeights<dim, spacedim>::reinit(
+    const hp::DoFHandler<dim, spacedim> &dof_handler,
+    const typename CellWeights<dim, spacedim>::WeightingFunction
+      &weighting_function)
   {
-    weighting_function =
-      [factor](const FiniteElement<dim, spacedim> &,
-               const typename hp::DoFHandler<dim, spacedim>::cell_iterator &)
-      -> unsigned int { return factor; };
+    tria_listener.disconnect();
+    tria_listener = dof_handler.get_triangulation().signals.cell_weight.connect(
+      make_weighting_callback(dof_handler, weighting_function));
   }
 
 
+
+  // ---------- static member functions ----------
+
+  // ---------- selection of weighting functions ----------
+
   template <int dim, int spacedim>
-  void
-  CellWeights<dim, spacedim>::register_ndofs_weighting(
-    const unsigned int factor)
+  typename CellWeights<dim, spacedim>::WeightingFunction
+  CellWeights<dim, spacedim>::constant_weighting(const unsigned int factor)
   {
-    weighting_function =
-      [factor](const FiniteElement<dim, spacedim> &active_fe,
-               const typename hp::DoFHandler<dim, spacedim>::cell_iterator &)
-      -> unsigned int { return factor * active_fe.dofs_per_cell; };
+    return
+      [factor](const typename hp::DoFHandler<dim, spacedim>::cell_iterator &,
+               const FiniteElement<dim, spacedim> &) -> unsigned int {
+        return factor;
+      };
   }
 
 
+
   template <int dim, int spacedim>
-  void
-  CellWeights<dim, spacedim>::register_ndofs_squared_weighting(
-    const unsigned int factor)
+  template <typename FactorType, typename ExponentType>
+  typename CellWeights<dim, spacedim>::WeightingFunction
+  CellWeights<dim, spacedim>::ndofs_weighting(
+    const std::pair<FactorType, ExponentType> &coefficients)
   {
-    weighting_function =
-      [factor](const FiniteElement<dim, spacedim> &active_fe,
-               const typename hp::DoFHandler<dim, spacedim>::cell_iterator &)
-      -> unsigned int {
-      return factor * active_fe.dofs_per_cell * active_fe.dofs_per_cell;
-    };
+    if (std::is_same<ExponentType, unsigned int>::value)
```

Comment:
you could save yourself some work by just picking one type for `ExponentType` :-)